### PR TITLE
feat: batch close

### PR DESCRIPTION
**Pull request by tool3**
# Batch Close
provide the ability to manually close batches using an array of `batchIds`

### API
```
const {BatchClose} = require('@applitools/eyes-sdk-core')
async function someFunction() {
...
await BatchClose().setBatchIds(['123', '456']).close()
...
}
```

you can also provide a custom URL:
```
const {BatchClose} = require('@applitools/eyes-sdk-core')
async function someFunction() {
...
await BatchClose().setUrl('example.com').setBatchIds(['123', '456']).close()
...
}
```

### CODE REVIEW

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      .replyWithError({message: 'something went wrong', code: 500})
+    const [err] = await presult(closeBatch({batchIds: ['678'], serverUrl, apiKey}))
+    expect(err.message).to.equal('Error: something went wrong')
+  })
+
+  it('should setup configuration', async () => {
+    const [err] = await presult(closeBatch({batchIds, serverUrl, apiKey}))
+    expect(err).to.be.undefined
+    batchIds.forEach((batchId, index) => {
+      expect(scopes[index].basePath).to.equal(serverUrl)
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:38]: I understand you use `scopes` because `serverConnector.deleteBatchSessions` doesn't return anything, right?
Otherwise you could use the reply to communicate the original request parameters back and not need this scope.
Just a thought, but indeed nothing to do here.

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      expect(scopes[index].interceptors[0].path).to.include(batchId)
+      expect(scopes[index].interceptors[0].queries).to.eql({apiKey})
+    })
+  })
+
+  it('should call deleteBatchSession per batchId', async () => {
+    await closeBatch({batchIds, serverUrl, apiKey})
+    batchIds.forEach((batchId, index) => {
+      expect(scopes[index].interceptors[0].path).to.include(batchId[index])
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:47]: shouldn't it be `batchIds[index]`? How does this pass?

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    const [err] = await presult(closeBatch({batchIds: ['678'], serverUrl, apiKey}))
+    expect(err.message).to.equal('Error: something went wrong')
+  })
+
+  it('should setup configuration', async () => {
+    const [err] = await presult(closeBatch({batchIds, serverUrl, apiKey}))
+    expect(err).to.be.undefined
+    batchIds.forEach((batchId, index) => {
+      expect(scopes[index].basePath).to.equal(serverUrl)
+      expect(scopes[index].interceptors[0].path).to.include(batchId)
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:39]: maybe you can be more explicit and do ``.to.equal(`/api/sessions/batches/${batchId}/close/bypointerid`)``

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      expect(scopes[index].interceptors[0].queries).to.eql({apiKey})
+    })
+  })
+
+  it('should call deleteBatchSession per batchId', async () => {
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:44]: what does this test do that `should setup configuration` doesn't? I think it can be removed

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  it('should throw', async () => {
+    nock(serverUrl)
+      .delete(`/api/sessions/batches/678/close/bypointerid`)
+      .query({apiKey})
+      .replyWithError({message: 'something went wrong', code: 500})
+    const [err] = await presult(closeBatch({batchIds: ['678'], serverUrl, apiKey}))
+    expect(err.message).to.equal('Error: something went wrong')
+  })
+
+  it('should setup configuration', async () => {
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:34]: A better name would be "should send the correct close batch requests to the server"

**Review by amitzur — CHANGES_REQUESTED** [review]
There are changes to be made.
Also new requirements:
1. add `setApiKey` method to fluent API
2. add `setProxy` method to fluent API and `proxy` param to `closeBatch`. The type of the proxy argument shouldn't matter here. It just passes through into the `Configuration` object.

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js (hunk 1 above)
Comment (tool3) [reply packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:38]: yeah since it doesn't return anything I needed a way to verify it was done successfully.

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js (hunk 2 above)
Comment (tool3) [reply packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:47]: because I am using `.includes` (and so index 0 for string `123` also works)
indeed a warning sign - updating this to explicit string

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js (hunk 3 above)
Comment (tool3) [reply packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:39]: I did exactly that before reading this comment 😆

File: packages/eyes-sdk-core/lib/close/BatchClose.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+function makeBatchClose(closeBatch) {
+  return function BatchClose() {
+    this._batchIds = null
```
Comment (amitzur) [inline packages/eyes-sdk-core/lib/close/BatchClose.js:3]: Since this is not called with `new`, then this code currently pollutes `global`.
What you want is to initialize `const ret = {}` and return that.

File: packages/eyes-sdk-core/lib/close/closeBatch.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  apiKey = process.env.APPLITOOLS_API_KEY,
+  proxy,
+}) {
+  const getAgentId = () => ''
+  const {logger, configuration} = setupServerConfig({serverUrl, apiKey, proxy})
+  const serverConnector = new ServerConnector({logger, configuration, getAgentId})
+
+  const promises = batchIds.map(batchId => serverConnector.deleteBatchSessions(batchId))
+  const results = await Promise.all(promises.map(presult))
+  const err = results.find(([err]) => !!err)
```
Comment (amitzur) [inline packages/eyes-sdk-core/lib/close/closeBatch.js:18]: it's confusing to call the argument in the inner function `err` and also assign the return value to a variable `err`.
I suggest renaming one of them, since they are not referring to the same thing.

File: packages/eyes-sdk-core/lib/close/closeBatch.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  if (err) throw err[0]
```
Comment (amitzur) [inline packages/eyes-sdk-core/lib/close/closeBatch.js:19]: why `err[0]` and not `err`? How did the tests pass? Isn't this a bug?

File: packages/eyes-sdk-core/lib/close/closeBatch.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  const logger = new Logger(!!process.env.APPLITOOLS_SHOW_LOGS)
+  const configuration = new Configuration()
+  if (serverUrl) configuration.setServerUrl(serverUrl)
+  if (apiKey) configuration.setApiKey(apiKey)
+  if (proxy) configuration.setProxy(proxy)
+  return {configuration, logger}
+}
+
+async function closeBatch({
+  batchIds = [],
```
Comment (amitzur) [inline packages/eyes-sdk-core/lib/close/closeBatch.js]: Not good. If user doesn't set batch ID's in the fluent API then you will silently not do anything.
You need to remove the default value and also have a test for this case (using the fluent API).
Notice the nuance in the test: the test for `BatchClose` which doesn't set batch ID's should throw an error that no batch ID's are set. So the `closeBatch` function that you inject needs to throw if there are not `batchIds`.

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    nock(serverUrl)
+      .delete(`/api/sessions/batches/999/close/bypointerid`)
+      .query({apiKey})
+      .replyWithError({message: 'something went wrong', code: 500})
+    const [err] = await presult(closeBatch({batchIds: ['888', '999'], serverUrl, apiKey}))
+    expect(err.message).to.equal('something went wrong')
+  })
+
+  it('should send the correct close batch requests to the server', async () => {
+    scopes = []
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:38]: `scopes` is initialized here, so no need to use `let scopes` 30 lines above. Just do `const scopes = []`.
While we're at it, how about:
```
const scopes = batchIds.map(...)
```

File: packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
+const nock = require('nock')
+const closeBatch = require('../../../lib/close/closeBatch')
+const {expect} = require('chai')
+const {presult} = require('../../../lib/troubleshoot/utils')
+
+describe('closeBatch', () => {
+  let serverUrl, apiKey, batchIds, scopes
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/closeBatch.spec.js:7]: I think all these variables need to be defined in each test individually. Or, at least define them here with `const`, so it's clear they don't change. But right now the code looks like they are meant to be changed. And there's no reason to initialize their values in `beforeEach`.

**Review by amitzur — CHANGES_REQUESTED** [review]

File: packages/eyes-sdk-core/lib/close/closeBatch.js (hunk 9 above)
Comment (amitzur) [reply packages/eyes-sdk-core/lib/close/closeBatch.js]: Sorry, it's my mistake about the test. It should only verify that `closeBatch` was called with `undefined`.

File: packages/eyes-sdk-core/lib/close/closeBatch.js (hunk 8 above)
Comment (tool3) [reply packages/eyes-sdk-core/lib/close/closeBatch.js]: since the return type of `presult` is `[err, value]`

File: packages/eyes-sdk-core/lib/close/closeBatch.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+const Configuration = require('../config/Configuration')
+const Logger = require('../logging/Logger')
+const {presult} = require('../troubleshoot/utils')
+
+async function closeBatch({batchIds, serverUrl, apiKey, proxy}) {
+  if (!batchIds) throw new Error('no batchIds were set')
+  const serverConnector = new ServerConnector({
+    logger: new Logger(!!process.env.APPLITOOLS_SHOW_LOGS),
+    configuration: new Configuration({serverUrl, apiKey}),
+    proxy,
```
Comment (amitzur) [inline packages/eyes-sdk-core/lib/close/closeBatch.js:13]: proxy is an argument for `Configuration`, like `serverUrl`. I wonder if this can be tested...

File: packages/eyes-sdk-core/test/unit/close/BatchClose.spec.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  it('should set proxy', async () => {
+    let proxy = {
+      protocol: 'https',
+      host: 'localhost',
+      port: 1234,
+      isHttpOnly: false,
+      url: 'http://localhost:1234',
+    }
+    let expectedProxy = {}
+    const BatchClose = makeBatchClose(({proxy}) => Object.assign(expectedProxy, proxy))
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/BatchClose.spec.js:54]: Any special reason for doing `Object.assign` and not `expectedProxy = proxy`?
In any case, if it's like this, prefer `const` over `let`.

File: packages/eyes-sdk-core/test/unit/close/BatchClose.spec.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+      .setUrl(serverUrl)
+      .setBatchIds(batchIds)
+      .setApiKey(apiKey)
+      .setProxy(proxy)
+      .close()
+
+    expect(response).to.eql({serverUrl, batchIds, apiKey, proxy})
+  })
+
+  it('should throw if no batchIds were provided', async () => {
```
Comment (amitzur) [inline packages/eyes-sdk-core/test/unit/close/BatchClose.spec.js:91]: `BatchClose` doesn't throw on its own. And there's already a test for `should set batchIds` so maybe you want to add it to that or rename this test to `should set empty batchIds`?

**Comment by amitzur** [comment]
:clap: :clap: :clap: :clap:
